### PR TITLE
feat: Grove Update Strategy Improvements Design

### PR DESCRIPTION
**Pull request by tmonty12**
#### What type of PR is this?

/kind documentation

#### What this PR does / why we need it:

This PR creates a design proposal for allowing configurability over the default Grove RollingUpdate strategy for PodCliqueSets, PodCliqueScalingGroups and PodCliques.
- Introduced a `ReplicaRecreate` strategy at the PodCliqueSet level to atomtically recreate PCS replicas in the case where application level version compatibility is not possible. Also introduces the notion of `maxUnavailable`/`maxSurge` for `ReplicaRecreate`
- For the PCS `RollingUpdate` strategy, introduces `maxSurge`/`maxUnavailable` at the PC and PCSG levels. 
- Includes considerations for the following:
  - Gang Scheduling
  - Indice management when surging
  - Use cases and examples
  - Updates to webhook validation
  - Updates to APIs

### CODE REVIEW

File: docs/designs/update-strategy.md [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+- Replica selection: ascending ordinal (lowest index first)
+- Entire PCSG replica is deleted and recreated (all member PodCliques together)
+
+**Within PCS Replica - Standalone PodClique (PC) Updates:**
+
+- Updates one pod at a time (sequential)
+- Pod selection: oldest pod first (by creation timestamp)
+- Individual pods are deleted and recreated
+
+This default behavior provides safe, conservative updates but lacks user configurability. At the PCSG and standalone PC levels, the update corresponds to maxUnavailable 1 and maxSurge 0 where a singular old replica is deleted and new one is created.
```
Comment (Ronkahn21) [inline docs/designs/update-strategy.md:28]: PC -> PCLQ

File: docs/designs/update-strategy.md [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+
+```go
+type PodCliqueSetUpdateStrategy struct {
+    Type          PodCliqueSetUpdateStrategyType
+    RollingUpdate *PodCliqueSetRollingUpdateStrategy
+}
+
+type PodCliqueSetUpdateStrategyType string
+const (
+    RollingUpdate    // Update replicas sequentially
```
Comment (Ronkahn21) [inline docs/designs/update-strategy.md:149]: the name does not match the description,. it does not update replicas sequentially it update the replica components sequentially, better name should be ReplicaUpdate or somthing that is refering to replica

File: docs/designs/update-strategy.md [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+    ReplicaRecreate  // Delete/recreate entire replica
```
[the remark this answers is not in the document]
Comment (Ronkahn21) [reply docs/designs/update-strategy.md]: I saw you ignore MaxSurge and MaxUnavailable of lower component  when ReplicaRecreate but I am not sure why do you need both strategy?

[the remark this answers is not in the document]
Comment (Ronkahn21) [reply docs/designs/update-strategy.md]: could we represent both strategy with just MaxSurge and MaxUnavailable ?

File: docs/designs/update-strategy.md (hunk 1 above)
Comment (gflarity) [reply docs/designs/update-strategy.md:28]: We do this a lot, will be a habit hard to break: https://docs.google.com/document/d/11dGjQh--s7QXggTyBxrcAttl2JHt1tbUKEV_UxvxAp8/edit?tab=t.0#heading=h.s62odkdoge8e

File: docs/designs/update-strategy.md [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+## Use Cases
+
+### Incompatible Version Updates
+
+Consider a multinode disaggregated deployment with 2 PCS replicas, where each replica contains:
+
+- **Frontend** standalone PC: 3 replicas
+- **Prefill** PCSG: 2 replicas (prefill-leader PC: 1 replica, prefill-worker PC: 2 replicas)
+- **Decode** PCSG: 2 replicas (decode-leader PC: 1 replica, decode-worker PC: 2 replicas)
+
```
Comment (gflarity) [inline docs/designs/update-strategy.md:39]: Suggest you include the YAML as well, it helps those who "see the matrix" ;)

```yaml
apiVersion: grove.io/v1alpha1
kind: PodCliqueSet
metadata:
  name: multinode-disaggregated-inference
  namespace: default
spec:
  # 2 PCS replicas - each replica contains the full inference pipeline
  replicas: 2
  template:
    podcliques:
    # Frontend - standalone PodClique (not part of any PCSG)
    - name: frontend
      spec:
        roleName: frontend
        replicas: 3
        podSpec:
          containers:
          - name: frontend
            # ...

    # Prefill Leader - part of prefill PCSG
    - name: prefill-leader
      spec:
        roleName: prefill-leader
        replicas: 1
        podSpec:
          containers:
          - name: prefill-leader
          # ...


    # Prefill Worker - part of prefill PCSG
    - name: prefill-worker
      spec:
        roleName: prefill-worker
        replicas: 2
        podSpec:
          containers:
          - name: prefill-worker
          # ... 

    # Decode Leader - part of decode PCSG
    - name: decode-leader
      spec:
        roleName: decode-leader
        replicas: 1
        podSpec:
          containers:
          - name: decode-leader
          # ... 

    # Decode Worker - part of decode PCSG
    - name: decode-worker
      spec:
        roleName: decode-worker
        replicas: 2
        podSpec:
          containers:
          - name: decode-worker
          # ... 

    # PodCliqueScalingGroups - groups of cliques that scale together
    podCliqueScalingGroups:
    # Prefill PCSG: 2 replicas of (1 leader + 2 workers)
    - name: prefill
      cliqueNames: [prefill-leader, prefill-worker]
      replicas: 2

    # Decode PCSG: 2 replicas of (1 leader + 2 workers)
    - name: decode
      cliqueNames: [decode-leader, decode-worker]
      replicas: 2
```

File: docs/designs/update-strategy.md [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+- This mixed-version state causes communication failures, incorrect behavior, or crashes
+
+The default strategy assumes version compatibility during the update window. For incompatible updates, users need the ability to tear down the entire PCS replica atomically to prevent any cross-version communication within a replica.
+
+### Compatible Configuration Update
+
+Consider a multinode aggregated inference serving deployment with 2 PCS replicas, where each replica contains:
+
+- **Aggregated Workers** PCSG: 3 replicas (each replica includes inference workers with a frontend capable of tokenization that accepts OpenAI Chat Completion requests)
+
```
Comment (gflarity) [inline docs/designs/update-strategy.md:64]: So I think your example would like the following. 

Sorry to nitpick, I know this is a toy example, but in this situation you could also just 1 replica os 6 pcsg right?  Or 6 pcs with 1 pcsg.  Which makes the most sense?
  
```yaml
  apiVersion: grove.io/v1alpha1
  kind: PodCliqueSet
  metadata:
    name: multinode-aggregated-inference
    namespace: default
  spec:
    # 2 PCS replicas - each replica contains the full aggregated inference deployment
    replicas: 2
    template:
      podcliques:
      # Aggregated Worker - combines inference + frontend/tokenization in a single pod
      # Each worker can handle OpenAI Chat Completion requests directly
      - name: aggregated-worker
        spec:
          roleName: aggregated-worker
          replicas: 1
          podSpec: #... 

      # PodCliqueScalingGroup - groups the aggregated workers that scale together
      podCliqueScalingGroups:
      # Aggregated Workers PCSG: 3 replicas of aggregated inference workers
      - name: aggregated-workers
        cliqueNames: [aggregated-worker]
        replicas: 3
```

File: docs/designs/update-strategy.md [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+- Partition-based rolling updates or canary deployments - These are advanced features that can be considered in future phases
+
+- Automatic strategy selection based on version compatibility detection - Users must explicitly configure update strategies
+
+- Automatic selection of surge based on capacity - Users are expected to be aware of their cluster capacity and should explicitly configure surge settings. Grove does not have visibility into available cluster capacity at update time and cannot automatically determine if surge is feasible
+
+## Proposal
+
+This proposal expands Grove's default RollingUpdate strategy to support replica recreation at the PodCliqueSet level for version incompatible upgrades, as well as `maxUnavailable` and `maxSurge` configuration at the PCS level. Additionally, it introduces `maxUnavailable` and `maxSurge` concepts at the standalone PodClique and PodCliqueScalingGroup levels to enable faster rollouts and capacity maintenance during compatible upgrades.
```
Comment (gflarity) [inline docs/designs/update-strategy.md]: Aside from added/work complexity, why not maxUnavailable and maxSurge at the PCS level as well?

File: docs/designs/update-strategy.md [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+
+## Architecture
+
+### Three-Level Update Control
+
+```
+PodCliqueSet (Top Level)
+├─ UpdateStrategy (controls PCS replica updates)
+│  ├─ RollingUpdate: one replica at a time
```
Comment (gflarity) [inline docs/designs/update-strategy.md:118]: I'm a bit confused by this. Even if we were to ReplicaRecreate, you'd still do it one at time as there's no maxSurge or maxUnavailable on PCS?

Comment (gflarity) [reply docs/designs/update-strategy.md:118]: Oh, I see below there is? This diagram is bit confusing.

File: docs/designs/update-strategy.md (hunk 2 above)
Comment (gflarity) [reply docs/designs/update-strategy.md:149]: Agreed,  ReplicaUpdate is better but it's still a bit confusing.  Not sure what a better name would be though. Unless we can find one,  we should document this very well.

File: docs/designs/update-strategy.md [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+```
+
+**Defaults:** `RollingUpdate` with `maxUnavailable=1, maxSurge=0`
+
+### 2. PodCliqueTemplateSpec.UpdateStrategy
+
+Controls **how pods update within a standalone PodClique**.
+
+```go
+type ComponentUpdateStrategy struct {
```
Comment (gflarity) [inline docs/designs/update-strategy.md:166]: Not sure the Component is necessary here. Just call it UpdateStrategy.  Though if recreate is set on the PCS, then this doesn't matter and we probably need to warn the user.

File: docs/designs/update-strategy.md [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+```
+
+**Defaults:** `maxUnavailable=1, maxSurge=0`
+**Scope:** Controls the update of PCSG replicas (groups of PodCliques)
+
+## Update Behavior
+
+### RollingUpdate (Default)
+
+The default RollingUpdate behavior is described in the [Motivation](#motivation) section. When using the RollingUpdate strategy, `maxUnavailable` and `maxSurge` settings at the PodCliqueSet level are invalid and will be rejected by the validation webhook - PCS replicas are always updated one at a time sequentially. However, PC and PCSG `updateStrategy` settings (maxUnavailable/maxSurge) are observed and control how pods and PCSG replicas update within each PCS replica.
```
Comment (gflarity) [inline docs/designs/update-strategy.md]: I'm bit confused by this section.  PCSGs always get recreated right now. We can use maxUnavailable/maxSurge to define how that happens more precisely.

File: docs/designs/update-strategy.md [hunk 10]
```diff
@@ -0,0 +1,661 @@
...
+    MaxSurge       *intstr.IntOrString  // How many extra PCSG replicas to create (absolute number or percentage)
+}
+```
+
+**Defaults:** `maxUnavailable=1, maxSurge=0`
+**Scope:** Controls the update of PCSG replicas (groups of PodCliques)
+
+## Update Behavior
+
+### RollingUpdate (Default)
```
Comment (gflarity) [inline docs/designs/update-strategy.md:191]: ReplicaUpdate

Comment (gflarity) [reply docs/designs/update-strategy.md:191]: Actually another way to Frame it DelegateChildren or something like that. IE follow the strategy of the children vs replica recreate which ignores their strategies and just recreates.

File: docs/designs/update-strategy.md [hunk 11]
```diff
@@ -0,0 +1,661 @@
...
+- Both PCS replicas can be deleted and recreated simultaneously
+- This provides faster updates but results in complete service unavailability during the recreation window
+
+**Use Cases:**
+
+- Version incompatible updates where old and new versions cannot coexist
+- Need to clear all state at once within a replica
+- Coordinated recreation of interdependent components to prevent cross-version communication issues
+
+## MaxSurge Considerations
```
Comment (gflarity) [inline docs/designs/update-strategy.md:234]: I found this section a bit hard to follow. Leaving some suggestions below to make it a bit easier.

File: docs/designs/update-strategy.md [hunk 12]
```diff
@@ -0,0 +1,661 @@
...
+When using ReplicaRecreate with `maxSurge > 0`, the update can get stuck if surge replicas fail to become available. This can happen in several scenarios:
+
+1. **Surge replica is unscheduled**: The surge replica's pods cannot be scheduled due to insufficient cluster resources, topology constraints that cannot be satisfied, node selectors/affinity mismatches, or resource quotas exceeded.
+
+2. **Surge replica has MinAvailable breached**: The surge replica's pods are scheduled but fail to become ready due to crash loops, health check failures, application startup failures, or dependency issues.
+
+3. **Existing replicas are unhealthy**: Even if surge replica is healthy, if existing replicas are unscheduled or have MinAvailable breached, the update may be blocked by `maxUnavailable` constraints.
+
+Users are responsible for identifying when a rolling update with `maxSurge` during ReplicaRecreate is stuck (e.g., update progress stalls, surge replica remains unscheduled or has MinAvailable breached) and manually intervening to unblock the update, such as by reducing `maxSurge` to 0 or deleting the stuck surge replica.
+
```
Comment (gflarity) [inline docs/designs/update-strategy.md:365]: I found this section a bit confusing. I worked with Opus 4.5 to make it a bit clearer.  Please take a look and incorporate what you like. 

```
# MaxSurge Considerations

## Overview

Enabling `maxSurge` changes the update strategy from **delete-then-create** (current default) to **create-then-delete**. This maintains full capacity during updates but introduces complexity around indexing, gang scheduling, and potential "stuck" scenarios.

**Key Risk**: With `maxSurge > 0`, updates can become **stuck** if surge resources fail to schedule or become healthy. Grove does not automatically detect or resolve these situations—users must monitor and manually intervene.


## Common Concepts Across All Levels

Before diving into level-specific details, here are concepts that apply at every level:

**Indexing Strategy**: Surge resources are assigned indices *above* the normal replica count to avoid index collisions. When old resources are deleted and recreated, they reclaim their original indices. Surge resources are deleted at the end of the update, leaving clean sequential indices.

**Availability Gating**: When `maxUnavailable=0`, the surge resource must become available *before* any old resource can be deleted. This is what enables zero-downtime updates but also creates the primary stuck scenario.


## PodClique MaxSurge (Pod Level)

**Scope**: Controls how individual pods update within a standalone PodClique.

**How It Works**:

- With `replicas=3` and `maxSurge=1`: surge pod gets index 3, update proceeds, surge pod deleted at end
- The index tracker fills holes from lowest to highest, ensuring no index gaps at completion

**Example with `replicas=3`, `maxSurge=1`, `maxUnavailable=0`:**

1. **Initial:** Pods with indices 0, 1, 2 (old spec)
2. **Create surge:** Pod with index 3 (new spec) — now have [0, 1, 2, 3]
3. **Delete pod 0:** Index 0 becomes available
4. **Recreate pod 0:** New pod fills index 0 (new spec) — now have [0, 1, 2, 3]
5. **Repeat for pods 1 and 2**
6. **Delete surge pod 3:** Final state [0, 1, 2] with new spec — no holes

**Gang Scheduling Impact**:

Surge pods are added to the **same PodGroup** as existing pods. This has an important implication:

> **Gang scheduling requires ALL pods in a PodGroup (including surge) to be schedulable together.**

If the cluster lacks resources for the surge pod, the entire gang becomes unschedulable, blocking the update.

**Stuck Scenarios**:

| Scenario | Cause | Result |
|----------|-------|--------|
| Surge pod unschedulable | Insufficient cluster resources | Gang blocked, update stuck |
| Surge pod not ready | Container failures, health check issues | Update blocked (if `maxUnavailable=0`) |


## PodCliqueScalingGroup MaxSurge (PCSG Replica Level)

**Scope**: Controls how PCSG replicas (groups of related PodCliques) update within a scaling group.

**How It Works**:

- With `replicas=3` and `maxSurge=1`: surge PCSG replica gets index 3
- Each PCSG replica contains multiple PodCliques that are updated together

**Example with `replicas=3`, `minAvailable=3`, `maxSurge=1`, `maxUnavailable=0`:**

1. **Initial:** Replicas 0, 1, 2 in base PodGang (old spec)
2. **Create surge:** Replica 3 in scaled PodGang (new spec)
3. **Wait for surge available:** Replica 3 becomes available
4. **Delete and recreate replicas 0, 1, 2** sequentially, waiting for each to become available
5. **Delete surge replica 3:** Final state replicas [0, 1, 2] with new spec

**Gang Scheduling Impact — The Base/Scaled PodGang Problem**:

This is where `maxSurge` becomes complicated. Grove uses a two-tier gang scheduling model:

- **Base PodGang**: Contains PCSG replicas 0 through `minAvailable-1`
- **Scaled PodGangs**: Contain PCSG replicas at index `minAvailable` and above

Since surge replicas are always at index ≥ `replicas` (which is ≥ `minAvailable`), **surge replicas always land in Scaled PodGangs**.

Scaled PodGangs have a dependency: they are **gated until the base PodGang is ready**. This creates a potential problem:

\`\`\`
┌─────────────────────────────────────────────────────────────────┐
│ POTENTIAL CIRCULAR DEPENDENCY                                   │
│                                                                 │
│ 1. Surge replica (in scaled PodGang) waits for base to be ready │
│ 2. Base PodGang is being updated (may not be "ready")           │
│ 3. Update needs surge to be available before deleting old base  │
│ 4. Deadlock: surge waits for base, update waits for surge       │
└─────────────────────────────────────────────────────────────────┘
\`\`\`

**Stuck Scenarios**:

| Scenario | Cause | Result |
|----------|-------|--------|
| Surge blocked by base PodGang | Base PodGang updating, not "ready" | Circular dependency, update stuck |
| Surge replica unschedulable | Resource constraints on scaled PodGang | Update stuck |
| Surge replica not available | Pod failures within the surge PCSG replica | Update blocked (if `maxUnavailable=0`) |
\```

## PCS Replica MaxSurge with ReplicaRecreate (Top Level)

**Scope**: Controls how entire PCS replicas are recreated during version-incompatible updates.

**How It Works**:

With `replicas=2`, `maxSurge=1`, and `maxUnavailable=0`:

```
Step 1: [0-old, 1-old]           Initial state
Step 2: [0-old, 1-old, 2-surge]  Create surge replica at index 2
Step 3: Wait for replica 2 to become available
Step 4: [0-new, 1-old, 2-surge]  Delete/recreate replica 0
Step 5: Wait for replica 0 to become available
Step 6: [0-new, 1-new, 2-surge]  Delete/recreate replica 1
Step 7: Wait for replica 1 to become available
Step 8: [0-new, 1-new]           Delete surge replica 2
```

This approach maintains full capacity (2 available replicas) throughout the update.

**Stuck Scenarios**:

Since surge PCS replicas are complete deployments (with their own PodGangs, PCSGs, and PodCliques), they can fail to become available for many reasons:

| Scenario | Examples |
|----------|----------|
| **Surge replica unscheduled** | Insufficient resources, topology constraints unsatisfiable, node selector mismatches, quota exceeded |
| **Surge replica unhealthy** | Container crash loops, health check failures, application startup failures, dependency issues |
| **Existing replicas degraded** | If `maxUnavailable` constraint prevents progress due to unhealthy existing replicas |


## User Responsibilities

**Grove does not automatically recover from stuck surge scenarios.** Users are responsible for:

1. **Monitoring update progress** — Watch for updates that stall (surge replica remains unscheduled or unhealthy)
2. **Diagnosing the cause** — Check pod events, resource availability, and PodGang status
3. **Manual intervention** — Options include:
   - Reducing `maxSurge` to 0 to switch to delete-then-create
   - Manually deleting the stuck surge replica
   - Freeing cluster resources to allow scheduling
   - Fixing application issues preventing readiness

## Summary

| Level | Surge Resource | Primary Risk | Gang Scheduling Concern |
|-------|---------------|--------------|------------------------|
| **PodClique** | Extra pod | Pod unschedulable | Surge pod blocks entire PodGroup gang |
| **PCSG** | Extra PCSG replica | Scaled PodGang gated | Base/scaled dependency creates circular wait |
| **PCS (ReplicaRecreate)** | Extra PCS replica | Replica unhealthy/unscheduled | Full replica must schedule and become healthy |

**Bottom Line**: `maxSurge` enables zero-downtime, full-capacity updates but shifts the failure mode from "reduced capacity during update" to "potentially stuck update requiring manual intervention."


```

Comment (gflarity) [reply docs/designs/update-strategy.md:365]: Please check out [the draft PR for rolling update E2E tests](https://github.com/ai-dynamo/grove/pull/280), seems like scaling during a roll is an edge case we care about current. What are the impacts of this on max surge? 

What if we just increase the replicas first (with the old revision), then treat this as a max unavailable situation once pods are ready.  Once the rollout is successful, reduce the replicas again. This might handle the intermingling mentioned above better. I suspect only really having to implement the maxUnavailable case would make the implementation a lot simpler too. The down side is you'll have to wait for new old pods to spin up, which could take quite a long time depending on what's being launched. That time would be wasted, but that might be a better trade off than all the stuck cases.

File: docs/designs/update-strategy.md (hunk 9 above)
Comment (gflarity) [inline docs/designs/update-strategy.md:193]:
```suggestion
The default ReplicaUpdate behavior is described in the [Motivation](#motivation) section. When using the ReplicaUpdate strategy, `maxUnavailable` and `maxSurge` settings at the PodCliqueSet level are invalid and will be rejected by the validation webhook - PCS replicas are always updated one at a time sequentially. However, PC and PCSG `updateStrategy` settings (maxUnavailable/maxSurge) are observed and control how pods and PCSG replicas update within each PCS replica.
```

**Comment by danbar2** [comment]
Stale PR with no comments for too long time
